# ND CSE Research: Care-Net

Alejandro Salvador Vega-Nogales
Advisor: Dr. Ronald Metoyer

## Project Background and Motivation (by Dr. Metoyer)

The Indiana 211 Partnership works with Indiana United Ways and its local subsidiaries to maintain an information database and provide a referral service. Annually, Indiana 211 reaches out to service providers to verify the accuracy of database records. Due to the overwhelming capacity requirements of this approach, over 1,100 organization updates are past due as of May 2019, and the oldest outdated record should have been updated in May 2018. Generally, accounts are expected to be removed if an organization does not respond within 90 days of missing the database record verification. However, Indiana 211 often keeps records active despite the lack of verification based on the importance of the services or the assumption that services are continued. This approach has lead to an abundance of obsolete and inaccurate data, wasting time and resources for populations in need and those who support them. This problem is compounded for those with complex needs that demand multiple services from various providers or for populations where the service network is highly informal and consisting of activists and volunteers. 

Nonetheless, identifying and accessing social services is critical in aiding vulnerable individuals and families in their move toward self-sufficiency; this is particularly true for individuals new to negotiating networks of services, including first-time domestic abuse victims, the recently unemployed, or the formerly incarcerated. This pilot project seeks to answer the question: How do you efficiently and effectively connect p
eople in need, as well as those helping them, to services and resources (informal and formal) available in a community that is continuously changing?

## Initial Data Exploration

Our dataset is very limited particularly in the short description that could be used to compare and differentiate different **Programs** and their **Providers**. 

# Data Source

1. ND_active.xslx: IN 211 data of Agencies (Programs), Services (Provides), and Sites (physical locations) of the social services. Mostly descriptions and metadata about the location or service

2. 

## Data Integration, Refactoring, and Save

In [0]:
    from google.colab import drive
    drive.mount('/content/drive/')
    prefix = './drive/My Drive/ND_CSE/Year_1/Research:Care-Net/JBDF_CareNet/Care-Net Backend/code_and_data'
    !echo "Project dir contents:" && ls -l "$prefix/data"
    !echo -e "\nColab Notebook home dir:" && ls

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).
Project dir contents:
total 32997082
-rw------- 1 root root    59862516 Feb 10 10:30  211_taxo_211_nodes.json
-rw------- 1 root root     2552771 Feb 10 10:30  211_taxonomy_edge_list.csv
-rw------- 1 root root          83 Feb 10 10:29  211_taxonomy_facets.csv
-rw------- 1 root root          72 Feb 10 10:29  211_taxonomy_kg_skel.json
-rw------- 1 root root    13684330 Feb 10 10:29  211_taxonomy_nodes.json
-rw------- 1 root root     1233445 Feb 10 10:30  211_taxonomy_term_references.csv
-rw------- 1 root root        7976 Oct 28 22:45  211_taxonomy_xml_schema.xsd
-rw------- 1 root root      365595 Feb 10 10:30  agency_id_to_node_num.json
-rw------- 1 root root   142992550 Dec 14 21:04  BANE_dense_feats.csv
-rw------- 1 root root   266426047 Dec 14 20:08  BANE_sparse_feats.json
-rw------- 1 root root        3758 Feb 10 10:38  code_to_name.json
-rw------- 1 root 

In [0]:
# install conda
install_conda = ''
while install_conda not in ('Y', 'N'):
  install_conda = input("Do you want to install conda?").upper()
install_conda = True if install_conda == 'Y' else False 
if install_conda:
  !wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && bash Miniconda3-4.5.4-Linux-x86_64.sh -bfp /usr/local

In [0]:
# !unset PYTHONPATH
# # import sys
# # sys.path.append('/usr/local/lib/python3.7/site-packages:$PATH" >> ~/.bashrc
# !echo "export PATH=/usr/local/lib/python3.7/site-packages:$PATH" >> /etc/bash.bashrc
# !echo "export PATH=/usr/local/lib/python3.7/site-packages:$PATH" >> /etc/bash.bash_profile
# !rm /etc/profile.d/conda.sh && ln -s /usr/local/etc/profile.d/conda.sh /etc/profile.d/conda.sh
# !echo ". /usr/local/etc/profile.d/conda.sh" >> ~/.bashrc

In [0]:
# !conda install -c conda-forge -y conda-pack

In [0]:
# !conda install --file './drive/My Drive/ND_CSE/Year_1/Research:Care-Net/code_and_data/binder/environment.yml'

In [0]:
# create_env = ''
# while create_env not in ('Y', 'N'):
#   create_env = input("Do you want to create the project env?").upper()
# create_env = True if create_env == 'Y' else False 
# if create_env:
#   # install the env and packages for our project
#   !conda env create -f './drive/My Drive/ND_CSE/Year_1/Research:Care-Net/code_and_data/binder/environment.yml'

In [0]:
# save_env = ''
# while save_env not in ('Y', 'N'):
#   save_env = input("Do you want to save the project conda env?").upper()
# save_env = True if save_env == 'Y' else False 
# if save_env:
#   !conda pack -n nd_care-net -j -1

In [0]:
# !conda install conda

/bin/bash: conda: command not found


In [0]:
# !source activate /usr/local/conda/envs/nd_care-net/activate
# !which python

In [0]:
# # activate and use the installed env
# !source activate
# !conda activate care-net

In [0]:
# !pip install pandas matplotlib keras 

In [0]:
import pandas as pd
import matplotlib as plt

In [0]:
data_path = prefix + '/data/ND_Active.xlsx'

# read the different sheets/tabs in the file
agency_df = pd.read_excel(data_path, sheet_name='Agency')
service_df = pd.read_excel(data_path, sheet_name='Service')
site_df = pd.read_excel(data_path, sheet_name='Site')

In [0]:
refs_df.describe()

129495  ...                                 Unnamed: 27
count   276318  ...                                           4
unique   63243  ...                                           2
top     116423  ...  InteractionReferral_ReferralsModule_amount
freq       154  ...                                           2

[4 rows x 28 columns]

In [0]:
# (naively) count the number of urls and how many are uniq

agency_urls = agency_df["AgencySystem_Websiteurl"]
service_urls = service_df["ServiceSystem_Websiteurl"]
site_urls = site_df["SiteSystem_Websiteurl"]

agcy_url_cnt = agency_urls.count()
a_uniq_cnt = agency_urls.nunique()
serv_url_cnt = service_urls.count()
se_uniq_cnt = service_urls.nunique()
site_url_cnt = site_urls.count()
si_uniq_cnt = site_urls.nunique()

entity_data = {
    'Agency': {'n_urls': agcy_url_cnt, 'uniq_urls': a_uniq_cnt},
    'Service': {'n_urls': serv_url_cnt, 'uniq_urls': se_uniq_cnt},
    'Site': {'n_urls': site_url_cnt, 'uniq_urls': si_uniq_cnt}
}

for t in entity_data:
    print("In the IN 211 snapshot, {entity} entities have {n_urls} of which {n_uniq} are unique".format(
        entity=t, n_urls=entity_data[t]['n_urls'], n_uniq=entity_data[t]['uniq_urls']))

In the IN 211 snapshot, Agency entities have 4905 of which 4166 are unique
In the IN 211 snapshot, Service entities have 4418 of which 2116 are unique
In the IN 211 snapshot, Site entities have 9319 of which 5046 are unique


In [0]:
agency_df.describe()

Agency_Id              AgencySystem_Name  \
count       6703                           6703   
unique      6703                           6661   
top         1347  First United Methodist Church   
freq           1                              6   

       AgencyCustom_Akakeywordsearchable  \
count                               2930   
unique                              2836   
top                   Highway Department   
freq                                  63   

                                 AgencySystem_Description  \
count                                                6690   
unique                                               3897   
top     An elected official responsible for administer...   
freq                                                 1006   

       AgencyOption_AgencyLegalStatus AgencyOption_IrsStatus  \
count                            6697                    608   
unique                              5                      8   
top                    ["Government"]          ["501(c)(3)"]   
freq                             3145                    585   

            AgencyHoursofoperation_ModuleHoursofoperation  \
count                                                4964   
unique                                               1366   
top     {"open":{"Friday":[[[8,0],[16,0]]],"Thursday":...   
freq                                                  716   

       AgencyCustom_NonStandardHours    AgencySystem_Websiteurl  \
count                           6627                       4905   
unique                          1659                       4166   
top                             Vary  www.jenningscounty-in.gov   
freq                             909                         21   

                AgencyCustom_Facebook          AgencyCustom_Twitter  \
count                             333                           120   
unique                            332                           119   
top     www.facebook.com/MyFranciscan  www.twitter.com/MyFranciscan   
freq                                2                             2   

       AgencyCustom_Phone1 AgencyCustom_Phone1Note  
count                 6607                     965  
unique                6397                     338  
top           260-668-1000            Headquarters  
freq                    11                      89

In [0]:
service_df.describe()

Agency_Id Service_Id  ServiceSystem_Name  \
count      16548      16548               16548   
unique      6703      16548                5628   
top         4278       2788  Trustee Assistance   
freq          63          1                1004   

                                    ServiceSystem_Aliases  \
count                                                3009   
unique                                               1013   
top     EAP, LIHEAP, LI-HEAP, Low Income Home Energy A...   
freq                                                   75   

                                ServiceSystem_Description  \
count                                               16548   
unique                                              10930   
top     May be able to assist township residents who a...   
freq                                                 1003   

                             ServiceSite_ServiceSite_Both  \
count                                               16497   
unique                                               9951   
top     [{"Site_Id":8831,"Name":"Vanderburgh County He...   
freq                                                   16   

       ServiceSite_ServiceSite ServiceGeotagus_ServiceGeotagusalternative  \
count                    16497                                      16548   
unique                    9951                                       2891   
top         [{"Site_Id":5487}]           [{"scope":"state","state":"IN"}]   
freq                        16                                       1157   

         ServiceTaxonomy_ModuleServicepost ServiceSystem_HoursOfOperation  \
count                                16548                          16407   
unique                                5614                           6598   
top     [{"taxonomy_code":"BD-1800.2000"}]                Mon-Fri 8am-4pm   
freq                                   380                           1269   

        ... ServiceCustom_DoesTheClientNeedToBringAnything  \
count   ...                                          13153   
unique  ...                                              3   
top     ...                                           true   
freq    ...                                           7303   

       ServiceSystem_DocumentsRequired ServiceOption_OptionlistWhatToBring  \
count                            12440                                  17   
unique                            2774                                   7   
top                     Nothing needed                        ["Photo ID"]   
freq                              1704                                   6   

       ServiceCustom_IsThereAFeeForTheService ServiceSystem_FeeStructure  \
count                                   15175                       7774   
unique                                      3                       2760   
top                                     false             Generally none   
freq                                     9786                        476   

       ServiceOption_MethodOfPaymentMultiSelect  \
count                                      2184   
unique                                      231   
top                   ["Cash","Personal Check"]   
freq                                        383   

       ServiceOption_LanguagesOtherThanEnglish  \
count                                     1879   
unique                                      30   
top                                ["Spanish"]   
freq                                      1290   

                          ServiceCustom_ProgramWaitPeriod  \
count                                                  36   
unique                                                 30   
top     Allow at least 15 business days after submitti...   
freq                                                    3   

       ServiceSystem_ServiceCapacityAndType  \
count                                     2   
unique                                    2   
top                     

In [0]:
site_df.describe()

Agency_Id Site_Id                      SiteSystem_Name  \
count      11521   11521                                11521   
unique      6703   11521                                11243   
top         2397   10205  Parkview - Firstcare Walk-In Clinic   
freq         130       1                                    9   

                                   SiteSystem_Description  \
count                                                8668   
unique                                               3397   
top     An elected official responsible for administer...   
freq                                                 1029   

              SiteHoursofoperation_ModuleHoursofoperation  \
count                                                6722   
unique                                               1982   
top     {"open":{"Monday":[[[8,0],[16,0]]],"Tuesday":[...   
freq                                                 1015   

       SiteSystem_HoursOfOperation SiteCustom_PublicTransportation  \
count                         8576                           11377   
unique                        2645                               3   
top                Mon-Fri 8am-4pm                           false   
freq                          1028                            9687   

       SiteOption_AccessForPeopleWithDisabilities  \
count                                        8032   
unique                                          5   
top                            ["ADA Accessible"]   
freq                                         7030   

       SiteCustom_SameAsMailingAddress SiteCustom_IsThisAddressConfidential  \
count                            11352                                11412   
unique                               3                                    3   
top                               true                                false   
freq                              9861                                11273   

        ... SiteAddressus_SiteAddressus_address_1  \
count   ...                                 10715   
unique  ...                                  9040   
top     ...            200 EAST WASHINGTON STREET   
freq    ...                                    24   

       SiteAddressus_SiteAddressus_address_2 SiteAddressus_SiteAddressus_city  \
count                                    932                            11012   
unique                                   583                             1076   
top                                  SUITE A                     INDIANAPOLIS   
freq                                      19                              832   

       SiteAddressus_SiteAddressus_county SiteAddressus_SiteAddressus_state  \
count                               10983                             10849   
unique                                294                                34   
top                                MARION                                IN   
freq                                  844                             10595   

       SiteAddressus_SiteAddressus_zip SiteAddressus_SiteAddressus_country  \
count                            11070                               11492   
unique                             916                                   3   
top                              46204                                  US   
freq                               212                               11478   

       SiteSystem_Websiteurl SiteCustom_Phone1 SiteCustom_Phone1Type  
count                   9319              8938                   715  
unique                  5046              7603                   360  
top           www.wic.in.gov    1-800-403-0864          Headquarters  
freq                     123               104                    44  

[4 rows x 21 columns]

In [0]:
from typing import List, Dict, Set, Optional
from tabulate import tabulate

# show dataset columns/features
def feat_table(df: pd.DataFrame, df_name: str, n_cols: int = 4) -> List[str]:
    # dataset features
    df_feats = list(df.columns)
    # table to hold n_cols feats per row
    feats_printable = []
    num_feats = len(df_feats)
    # go through feats n_cols at a time
    for i in range(0, num_feats, n_cols):
        if (i + n_cols) >= num_feats:
            empty_slots = n_cols - (num_feats - i)
            rem_feats = df_feats[i::] + (['[EMPTY]'] * empty_slots)
            row = ["{}-{}".format(i+1,i+n_cols)] + rem_feats
            feats_printable.append(row)
        else:
            row = ["{}-{}".format(i+1,i+n_cols)] + df_feats[i:i+n_cols]
            feats_printable.append(row)
    print("{name} dataset has following columns:\n {feat_table}".format(name=df_name, feat_table=tabulate(feats_printable)))
    return df_feats

In [0]:
# print more readable table of features and get list of feats
agency_feats = feat_table(agency_df, 'Agency', 4)

Agency dataset has following columns:
 -----  ------------------------------  ----------------------  ---------------------------------------------  -----------------------------
1-4    Agency_Id                       AgencySystem_Name       AgencyCustom_Akakeywordsearchable              AgencySystem_Description
5-8    AgencyOption_AgencyLegalStatus  AgencyOption_IrsStatus  AgencyHoursofoperation_ModuleHoursofoperation  AgencyCustom_NonStandardHours
9-12   AgencySystem_Websiteurl         AgencyCustom_Facebook   AgencyCustom_Twitter                           AgencyCustom_Phone1
13-16  AgencyCustom_Phone1Note         [EMPTY]                 [EMPTY]                                        [EMPTY]
-----  ------------------------------  ----------------------  ---------------------------------------------  -----------------------------


In [0]:
service_feats = feat_table(service_df, 'Service', 4)

Service dataset has following columns:
 -----  ----------------------------------------  ----------------------------------------------------  --------------------------------------  ----------------------------------------------
1-4    Agency_Id                                 Service_Id                                            ServiceSystem_Name                      ServiceSystem_Aliases
5-8    ServiceSystem_Description                 ServiceSite_ServiceSite_Both                          ServiceSite_ServiceSite                 ServiceGeotagus_ServiceGeotagusalternative
9-12   ServiceTaxonomy_ModuleServicepost         ServiceSystem_HoursOfOperation                        ServiceCustom_ClosingInformation        ServiceSystem_Websiteurl
13-16  ServiceSystem_EmailAddress                ServiceSystem_PhoneMain                               ServiceCustom_MatchlistPhone1Note       ServiceCustom_MatchlistPhone2
17-20  ServiceCustom_MatchlistPhone2Note         ServiceCustom_MatchlistPhone3

In [0]:
site_feats = feat_table(site_df, 'Site', 4)

Site dataset has following columns:
 -----  -------------------------------------------  ------------------------------------  ----------------------------------  ------------------------------------------
1-4    Agency_Id                                    Site_Id                               SiteSystem_Name                     SiteSystem_Description
5-8    SiteHoursofoperation_ModuleHoursofoperation  SiteSystem_HoursOfOperation           SiteCustom_PublicTransportation     SiteOption_AccessForPeopleWithDisabilities
9-12   SiteCustom_SameAsMailingAddress              SiteCustom_IsThisAddressConfidential  SiteAddressus_SiteAddressus_id      SiteAddressus_SiteAddressus_address_1
13-16  SiteAddressus_SiteAddressus_address_2        SiteAddressus_SiteAddressus_city      SiteAddressus_SiteAddressus_county  SiteAddressus_SiteAddressus_state
17-20  SiteAddressus_SiteAddressus_zip              SiteAddressus_SiteAddressus_country   SiteSystem_Websiteurl               SiteCustom_Phone1
21-24  S

In [0]:
# select cols we'll use
# NOTE: this is HARDCODED for the ND_Active.xslx file
ag_bitmap = list(map(lambda i: bool(i), [1,1,1,1,0,0,1,1,1,1,1,1,1]))
serv_bitmap = list(map(lambda i: bool(i), [1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]))
site_bitmap = list(map(lambda i: bool(i), [1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,0,1,1,0]))            

def feats_from_bitmap(feat_bitmap: List[bool], df_name: str, feats: List[str]) -> None:
    print("Selected raw features from {} dataset are:".format(df_name))
    for idx, bit in enumerate(feat_bitmap):
        if bit:
            print(feats[idx])
    print("---"*20, "\n"*2)

In [0]:
feats_from_bitmap(ag_bitmap, 'Agency', agency_feats)

Selected raw features from Agency dataset are:
Agency_Id
AgencySystem_Name
AgencyCustom_Akakeywordsearchable
AgencySystem_Description
AgencyHoursofoperation_ModuleHoursofoperation
AgencyCustom_NonStandardHours
AgencySystem_Websiteurl
AgencyCustom_Facebook
AgencyCustom_Twitter
AgencyCustom_Phone1
AgencyCustom_Phone1Note
------------------------------------------------------------ 




### Agency new Features Description:
* Agency_Id -> **AGENCY_ID**: int = UID per Agency
* AgencySystem_Name -> **AGENCY_NAME** : str = self-descriptive
* AgencyCustom_Akakeywordsearchable : str (comma sep.) -> **AGENCY_AKA_KEYWORDS** : List\[str\] = List of relevant keywords or Agency aliases
* AgencySystem_Description -> **AGENCY_DESCR** : str = ~1-2 sentences of plain text agency description
* AgencyHoursofoperation_ModuleHoursofoperation : Dict ( {'open': {<Weekday\>: \[ \[ \[start_H, start_Min], \[end_H, end_Min] ] ],...}, 'closed': { } } )  
    -> **AGENCY_HRS** : Dict ( {'open': {<Weekday\>: {'start': [ [start_H, start_Min], ... ], 'end': [ [end_H, end_Min], ... ], 'closed': { } }  = Agency Hours of Operation
    **NOTE**: Would need to eval each entry to parse as Dict
* AgencyCustom_NonStandardHours -> **AGENCY_TXT_HRS** : str = Agency Hours of Operation in Plain Text (eg Mon-Fri 8am-5pm)
* AgencySystem_Websiteurl -> **AGENCY_WEB_URL** : Optional[str] = Agency's website
* AgencyCustom_Facebook -> **AGENCY_FB_URL** : Optional[str] = Agency's FB page 
* AgencyCustom_Twitter -> **AGENCY_TW_URL** : Optional[str] = Agency's Twitter page
* AgencyCustom_Phone1 -> **AGENCY_PHONE** : str = Agency's phone number
* AgencyCustom_Phone1Note -> **AGENCY_PHONE_NOTE** : Optional[str] = A note about an Agency's Phone Number (eg dept, type of number (tollfree), etc)

**NOTE**: datatypes are the same except explicitly stated next to OG feature name

In [0]:
feats_from_bitmap(serv_bitmap, 'Service', service_feats)

Selected raw features from Service dataset are:
Agency_Id
Service_Id
ServiceSystem_Name
ServiceSystem_Aliases
ServiceSystem_Description
ServiceSite_ServiceSite_Both
ServiceGeotagus_ServiceGeotagusalternative
ServiceTaxonomy_ModuleServicepost
ServiceSystem_HoursOfOperation
ServiceCustom_ClosingInformation
ServiceSystem_Websiteurl
ServiceSystem_EmailAddress
ServiceSystem_PhoneMain
ServiceCustom_MatchlistPhone1Note
ServiceCustom_MatchlistPhone2
ServiceCustom_MatchlistPhone2Note
ServiceCustom_MatchlistPhone3
ServiceCustom_MatchlistPhone3Note
ServiceSystem_Fax
ServiceSystem_EligibilityTextArea
ServiceCustom_AreThereAnyPrerequisitesForThisService
ServiceSystem_ApplicationintakeProcess
ServiceCustom_DoesTheClientNeedToBringAnything
ServiceSystem_DocumentsRequired
ServiceOption_OptionlistWhatToBring
ServiceCustom_IsThereAFeeForTheService
ServiceSystem_FeeStructure
ServiceOption_MethodOfPaymentMultiSelect
ServiceOption_LanguagesOtherThanEnglish
ServiceCustom_ProgramWaitPeriod
ServiceSystem_Serv

### Service new Features Description: 
* Agency_Id -> **AGENCY_ID**: int = UID for corresponding Agency
* Service_Id -> **SERV_ID**: int = UID per Service
* ServiceSystem_Name -> **SERV_NAME**: str = self-descriptive
* ServiceSystem_Aliases -> **SERV_AKA**: str (comma sep) = Service aliases
* ServiceSystem_Description -> **SERV_DESCR**: str = ~1-2 sentences of plain text Service description
* ServiceSite_ServiceSite_Both -> **SERV_SITES**: List\[Dict] ( [ {'Site_Id': <int>, 'Name': <str>}, ...] = The site(s) where a service is offered  
**NOTE**: if AGENCY_ID && SERV_NAME are the same for several services we could join the rows to avoid duplicate results with different locations
* ServiceGeotagus_ServiceGeotagusalternative -> **SERV_GEO_TAG** : List[Dict] (\[ {'scope' : <state, county, city, zip\>, ... (keys from broadest to more specific starting from scope) } ]  
**NOTE**: There are multiple locations per service in many cases meaning data conflicts in treating them separately or jointly w. diff. locs.
* ServiceTaxonomy_ModuleServicepost -> **SERV_TAXONOMY_CODES** : List[Dict] (\[ {'taxonomy_code': str}, ...]) = The relevant codes in the [211 LA County Taxonomy of Human Services](https://www.211la.org/node/305)
* ServiceSystem_HoursOfOperation -> **SERV_OP_HOURS** : str (semicolon sep.) = Plain text Description of Operating hours (some indicate type (eg office, campus, answering, etc))
* ServiceCustom_ClosingInformation -> **SERV_CLOSED_TIMES** : Optional[str] = Plain text description of times the service is closed (eg state/federal holidays, closed on winter holidays, etc)
* ServiceSystem_Websiteurl -> **SERV_WEB_URL** : Optional[str] = Service's website
* ServiceSystem_EmailAddress -> **SERV_EMAIL** : Optional[str] = Service's contact email
* ServiceSystem_PhoneMain -> **SERV_PHONE_1** : Optional[str] = Service's main phone number
* ServiceCustom_MatchlistPhone1Note -> **SERV_PHONE_1_NOTE** : Optional[str] = A note about a Service's main phone number
* ServiceCustom_MatchlistPhone2 -> **SERV_PHONE_2** : Optional[str] = Service's 2nd phone number
* ServiceCustom_MatchlistPhone2Note -> **SERV_PHONE_2_NOTE** : Optional[str] = A note about a Service's 2nd phone number
* ServiceCustom_MatchlistPhone3 -> **SERV_PHONE_3** : Optional[str] = Service's 3rd phone number
* ServiceCustom_MatchlistPhone3Note -> **SERV_PHONE_3_NOTE** : Optional[str] = A note about a Service's 3rd phone number
* ServiceSystem_Fax -> **SERV_FAX** : Optional[str] = Service fax number
* ServiceSystem_EligibilityTextArea -> **SERV_ELIGIBLE_DESCR** : str = A plaintext description of conditions for being eligible for the Service
* ServiceCustom_AreThereAnyPrerequisitesForThisService -> **SERV_HAS_PRE_REQ** : Optional[bool] = T/F/None for Service prerequisites
* ServiceSystem_ApplicationintakeProcess -> **SERV_INTAKE_PROC** : str = A plaintext description of the process for applying for service or making an appointment
* ServiceCustom_DoesTheClientNeedToBringAnything -> **SERV_CLIENT_NEEDS_TO_BRING** : Optional[bool] = T/F/None if the client needs to bring anything
* ServiceSystem_DocumentsRequired -> **SERV_REQ_DOCS** : str (asterisk sep, semicolon sep, and others...) = A plaintext description of the documents, and other proof required for the Service
* ServiceOption_OptionlistWhatToBring -> **SERV_OPTS_BRING_LIST** : Optional[List[int]] = List of codes for documents clients need to bring ([1458 => "Copy of birth certificate", 1457 => "Photo ID", 1459 => "Program vouchers", 1460 -> "Proof of address", 1461 => "Proof of income and expense documentation", 1462 => "Proof of insurance", 1463 => "Social Security Card"])  
**NOTE**: this is very useful but none or little of the data seems to have this attr
* ServiceCustom_IsThereAFeeForTheService -> **SERV_HAS_FEE** : Optional[bool] = T/F/None if Service requires a fee
* ServiceSystem_FeeStructure -> **SERV_FEE_STRUCT** : str = Breaks down costs of Services provided (but most just say 'varies')
* ServiceOption_MethodOfPaymentMultiSelect -> **SERV_PAY_METHODS** : List[str] = List of accepted payment methods for Services (eg cash, checks, etc)  
**NOTE**: These codes exist [324 => "Accepts cash, credit card, or personal check", 325 => "Accepts Medicare, Medicaid and other third party insurance", 326 => "Call for details", 328 => "Not applicable", 861 => "Cash", 862 => "Personal Check", 863 => "Credit Card", 864 => "Debit Card", 327 => "Private Insurance", 865 => "EBT Card", 866 => "Medicaid", 867 => "Medicaid Wavier", 868 => "Medicare", 869 => "Healthy Indiana Plan (HIP)", 1110 => "PEPW", 1292 => "Money Order", 1449 => "Cashiers Check", 2235 => "CHOICE Voucher", 2236 => "Veterans Benefits", 2237 => "SNAP Double Bucks"] but most are just empty (not even text)
* ServiceOption_LanguagesOtherThanEnglish -> **SERV_ALT_LANGS** : List[str] = List of languages other than English spoken at that Service
* ServiceCustom_ProgramWaitPeriod -> **SERV_WAIT_PERIOD** : str = Plaintext description of the waiting period for a Service (**NOTE**: most are empty)
* ServiceSystem_ServiceCapacityAndType -> **SERV_CAP_AND_TYPE** : str = not sure. not a single example in small data. need to check complete data
* ServiceCustom_HelplineNotes -> **SERV_HELPLINE_NOTES** : str = not sure. need to check complete data

In [0]:
feats_from_bitmap(site_bitmap, 'Site', site_feats)

Selected raw features from Site dataset are:
Agency_Id
Site_Id
SiteSystem_Name
SiteSystem_Description
SiteHoursofoperation_ModuleHoursofoperation
SiteSystem_HoursOfOperation
SiteCustom_PublicTransportation
SiteOption_AccessForPeopleWithDisabilities
SiteCustom_SameAsMailingAddress
SiteCustom_IsThisAddressConfidential
SiteAddressus_SiteAddressus_address_1
SiteAddressus_SiteAddressus_address_2
SiteAddressus_SiteAddressus_city
SiteAddressus_SiteAddressus_state
SiteAddressus_SiteAddressus_zip
SiteSystem_Websiteurl
SiteCustom_Phone1
------------------------------------------------------------ 




### Site New Features Description: 
* Agency_Id -> **AGENCY_ID** : str = matching agency for a Site
* Site_Id -> **SITE_ID** : str = UID per Site
* SiteSystem_Name -> **SITE_NAME** : str = self-descriptive
* SiteSystem_Description -> **SITE_DESCR** : str = A plaintext description of a Site
* SiteHoursofoperation_ModuleHoursofoperation : Dict ( {'open': {<Weekday\>: \[ \[ \[start_H, start_Min], \[end_H, end_Min] ] ],...}, 'closed': { } } )  
    -> **SITE_HOURS** : Dict ( {'open': {<Weekday\>: {'start': [ [start_H, start_Min], ... ], 'end': [ [end_H, end_Min], ... ], 'closed': { } }  = Agency Hours of Operation
    **NOTE**: Would need to eval each entry to parse as Dict
* SiteSystem_HoursOfOperation -> **SITE_TXT_HOURS** : str = Plaintext descriptions of a Site's operating hours
* SiteCustom_PublicTransportation -> **SITE_HAS_PUBL_TRANSPO** : bool = T/F if Site is just next to (or near to???) public transportation
* SiteOption_AccessForPeopleWithDisabilities -> **SITE_HAS_DISABL_ACCESS** : List[str] = List of accessibility descriptions for a Site(eg ADA, not ADA, N/A, etc) (uses these codes but most ignore: [1112 => "ADA Accessible", 1113 => "Not ADA accessible but will make arrangements", 1114 => "Not ADA accessible", 2053 => "Not applicable"])
* SiteCustom_SameAsMailingAddress -> **SITE_ADDR_IS_MAIL** : Optional[bool] = T/F/None if Site's address is the same as their mailing address
* SiteCustom_IsThisAddressConfidential -> **SITE_ADDR_IS_CONF** : Optional[bool] = T/F/None if Site's Address should not be shared/disclosed
* SiteAddressus_SiteAddressus_address_1 
* SiteAddressus_SiteAddressus_address_2
* SiteAddressus_SiteAddressus_city
* SiteAddressus_SiteAddressus_state
* SiteAddressus_SiteAddressus_zip
(Previous 5 combined into one) --> **SITE_ADDR** : Dict ( { 'address1' : <SiteAddressus_SiteAddressus_address_1\>, 'address2': <SiteAddressus_SiteAddressus_address_2\>, 'city': <SiteAddressus_SiteAddressus_city\>, 'state': <SiteAddressus_SiteAddressus_state\>, 'zip': <SiteAddressus_SiteAddressus_zip\>}) = A Site's full address
* SiteSystem_Websiteurl -> **SITE_WEB_URL** : Optional[str] -> A Site's website
* SiteCustom_Phone1 -> **SITE_PHONE** : str = A Site's phone number

In [0]:
# delete unwanted cols/feats from dataframes

from itertools import compress

def cols_to_del(feats: List[str], bitmap: List[bool]) -> List[str]:
    # assuming bitmap where 1 means keep, 0 means del. negate bitmap
    rev_bm = list(map(lambda b: not b, bitmap))
    return list(compress(feats, rev_bm))
    
    
# refactor agency df
new_ag_feats = [
    'AGENCY_ID', 'name', 'AGENCY_AKA_KEYWORDS', 'description', 'AGENCY_HRS', 'AGENCY_TXT_HRS',
    'AGENCY_WEB_URL', 'AGENCY_FB_URL', 'AGENCY_TW_URL', 'AGENCY_PHONE', 'AGENCY_PHONE_NOTE'
]

# if the current columns don't match the new ones
# delete unwanted cols, and rename cols
if list(agency_df.columns) != new_ag_feats:
    ag_del_cols = cols_to_del(agency_feats, ag_bitmap)
    agency_df.drop(ag_del_cols, inplace=True, axis=1)
    agency_df.columns = new_ag_feats

# refactor service df
new_serv_feats = [
    'AGENCY_ID', 'SERV_ID', 'name', 'SERV_AKA', 'description', 'SERV_SITES', 'SERV_GEO_TAG', 
    'SERV_TAXONOMY_CODES', 'SERV_OP_HOURS', 'SERV_CLOSED_TIMES', 'SERV_WEB_URL', 'SERV_EMAIL', 'SERV_PHONE_1',
    'SERV_PHONE_1_NOTE', 'SERV_PHONE_2', 'SERV_PHONE_2_NOTE', 'SERV_PHONE_3', 'SERV_PHONE_3_NOTE', 'SERV_FAX',
    'SERV_ELIGIBLE_DESCR', 'SERV_HAS_PRE_REQ', 'SERV_INTAKE_PROC', 'SERV_CLIENT_NEEDS_TO_BRING', 'SERV_REQ_DOCS',
    'SERV_OPTS_BRING_LIST', 'SERV_HAS_FEE', 'SERV_FEE_STRUCT', 'SERV_PAY_METHODS', 'SERV_ALT_LANGS', 'SERV_WAIT_PERIOD', 
    'SERV_CAP_AND_TYPE', 'SERV_HELPLINE_NOTES'
]

# if the current columns don't match the new ones
# delete unwanted cols, and rename cols
if list(service_df.columns) != new_serv_feats:
    serv_del_cols = cols_to_del(service_feats, serv_bitmap)
    service_df.drop(serv_del_cols, inplace=True, axis=1)
    service_df.columns = new_serv_feats

# refactor site df

new_site_feats = [
    'AGENCY_ID', 'SITE_ID', 'name', 'description', 'SITE_HOURS', 'SITE_TXT_HOURS', 'SITE_HAS_PUBL_TRANSPO',
    'SITE_HAS_DISABL_ACCESS', 'SITE_ADDR_IS_MAIL', 'SITE_ADDR_IS_CONF', 'SITE_WEB_URL', 'SITE_PHONE', 'SITE_ADDR'
]

# if the current columns don't match the new ones
# combine the address feats into a single feature that holds all feats
# as a single dict/json
# delete the address feats, delete unwanted cols,
# add new address feat, and rename cols
if list(site_df.columns) != new_site_feats:
    # feats that need to be parsed and combined into a single SITE_ADDR feat
    comb_feats = [
        'SiteAddressus_SiteAddressus_address_1', 'SiteAddressus_SiteAddressus_address_2', 
        'SiteAddressus_SiteAddressus_city', 'SiteAddressus_SiteAddressus_state', 'SiteAddressus_SiteAddressus_zip' 
    ]
    # translate curr feat names to their new dict/json keys
    feat_trans = {
        'SiteAddressus_SiteAddressus_address_1': 'address1', 'SiteAddressus_SiteAddressus_address_2': 'address2', 
        'SiteAddressus_SiteAddressus_city': 'city', 'SiteAddressus_SiteAddressus_state': 'state', 
        'SiteAddressus_SiteAddressus_zip': 'zip' 
    }
    
    # subset of our df with just the address feats
    addr_df = site_df[comb_feats]
    addr_list = addr_df.to_dict(orient='records')
    del addr_df
    
    for i, addr in enumerate(addr_list):
        trans_addr = {}

        for k in addr.keys():
            new_key = feat_trans[k]
            trans_addr[new_key] = addr[k]
        addr_list[i] = trans_addr
        
    site_del_cols = cols_to_del(site_feats, site_bitmap)
    # delete unwanted cols and old address cols
    site_df.drop(site_del_cols, inplace=True, axis=1)
    site_df.drop(comb_feats, inplace=True, axis=1)
    
    # add new single address feat
    site_df['SITE_ADDR'] = addr_list
    site_df.columns = new_site_feats
    
    

In [0]:
agency_df.describe()

AGENCY_ID                           name AGENCY_AKA_KEYWORDS  \
count       6703                           6703                2930   
unique      6703                           6661                2836   
top         1347  First United Methodist Church  Highway Department   
freq           1                              6                  63   

                                              description  \
count                                                6690   
unique                                               3897   
top     An elected official responsible for administer...   
freq                                                 1006   

                                               AGENCY_HRS AGENCY_TXT_HRS  \
count                                                4964           6627   
unique                                               1366           1659   
top     {"open":{"Friday":[[[8,0],[16,0]]],"Thursday":...           Vary   
freq                                                  716            909   

                   AGENCY_WEB_URL                  AGENCY_FB_URL  \
count                        4905                            333   
unique                       4166                            332   
top     www.jenningscounty-in.gov  www.facebook.com/MyFranciscan   
freq                           21                              2   

                       AGENCY_TW_URL  AGENCY_PHONE AGENCY_PHONE_NOTE  
count                            120          6607               965  
unique                           119          6397               338  
top     www.twitter.com/MyFranciscan  260-668-1000      Headquarters  
freq                               2            11                89

In [0]:
agency_df.columns

Index(['AGENCY_ID', 'name', 'AGENCY_AKA_KEYWORDS', 'description', 'AGENCY_HRS',
       'AGENCY_TXT_HRS', 'AGENCY_WEB_URL', 'AGENCY_FB_URL', 'AGENCY_TW_URL',
       'AGENCY_PHONE', 'AGENCY_PHONE_NOTE'],
      dtype='object')

In [0]:
service_df.describe()

AGENCY_ID SERV_ID                name  \
count      16548   16548               16548   
unique      6703   16548                5628   
top         4278    2788  Trustee Assistance   
freq          63       1                1004   

                                                 SERV_AKA  \
count                                                3009   
unique                                               1013   
top     EAP, LIHEAP, LI-HEAP, Low Income Home Energy A...   
freq                                                   75   

                                              description  \
count                                               16548   
unique                                              10930   
top     May be able to assist township residents who a...   
freq                                                 1003   

                                               SERV_SITES  \
count                                               16497   
unique                                               9951   
top     [{"Site_Id":8831,"Name":"Vanderburgh County He...   
freq                                                   16   

                            SERV_GEO_TAG                 SERV_TAXONOMY_CODES  \
count                              16548                               16548   
unique                              2891                                5614   
top     [{"scope":"state","state":"IN"}]  [{"taxonomy_code":"BD-1800.2000"}]   
freq                                1157                                 380   

          SERV_OP_HOURS                  SERV_CLOSED_TIMES  ...  \
count             16407                               4400  ...   
unique             6598                                409  ...   
top     Mon-Fri 8am-4pm  CLOSED state and federal holidays  ...   
freq               1269                               1197  ...   

       SERV_CLIENT_NEEDS_TO_BRING   SERV_REQ_DOCS SERV_OPTS_BRING_LIST  \
count                       13153           12440                   17   
unique                          3            2774                    7   
top                          true  Nothing needed         ["Photo ID"]   
freq                         7303            1704                    6   

       SERV_HAS_FEE SERV_FEE_STRUCT           SERV_PAY_METHODS SERV_ALT_LANGS  \
count         15175            7774                       2184           1879   
unique            3            2760                        231             30   
top           false  Generally none  ["Cash","Personal Check"]    ["Spanish"]   
freq           9786             476                        383           1290   

                                         SERV_WAIT_PERIOD SERV_CAP_AND_TYPE  \
count                                                  36                 2   
unique                                                 30                 2   
top     Allow at least 15 business days after submitti...              text   
freq                                                    3                 1   

                                      SERV_HELPLINE_NOTES  
count                                                2542  
unique                                                872  
top     Specialists: The services listed above are exa...  
freq                                                  969  

[4 rows x 32 columns]

In [0]:
# drop first row of each df since it just describes the data types and structure
agency_df = agency_df[1::]
service_df = service_df[1::]
site_df = site_df[1::]

In [0]:
# can't be described because of dict data
site_df.head()

AGENCY_ID SITE_ID                                  name  \
1         1       1  Tippecanoe Valley School Corporation   
2         2       2                   Guidance Ministries   
3         3       3             Elkhart County Prosecutor   
4         4       4           United Way of Fulton County   
5         5       5       Fulton County Health Department   

                                         description  \
1  Administers the elementary, middle, and high s...   
2                                                NaN   
3  Represents the state of Indiana in felony and ...   
4                                                NaN   
5  A government agency that provides public healt...   

                                          SITE_HOURS  \
1  {"open":{"Monday":[[[7,30],[16,30]]],"Tuesday"...   
2  {"open":{"Tuesday":[[[9,0],[18,0]]],"Thursday"...   
3                                                NaN   
4  {"open":{"Friday":[[[10,0],[15,0]]],"Monday":[...   
5                                                NaN   

                                      SITE_TXT_HOURS SITE_HAS_PUBL_TRANSPO  \
1                              Mon-Fri 7:30am-4:30pm                 false   
2  Tues and Thurs 9am-6pm; Mon, Wed, and Fri 9am-4pm                 false   
3  Mon 8am-noon and 1pm-5pm; Tues-Fri 8am-noon an...                 false   
4                                   Mon-Fri 10am-3pm                 false   
5                                                NaN                 false   

  SITE_HAS_DISABL_ACCESS SITE_ADDR_IS_MAIL SITE_ADDR_IS_CONF  \
1     ["ADA Accessible"]              true             false   
2                    NaN             false             false   
3     ["ADA Accessible"]              true             false   
4     ["ADA Accessible"]              true             false   
5                    NaN              true             false   

                      SITE_WEB_URL    SITE_PHONE  \
1               www.tvsc.k12.in.us  574-598-2759   
2       www.guidanceministries.com  574-296-7192   
3  www.elkhartcountyprosecutor.com           NaN   
4                              NaN  574-223-8929   
5              www.co.fulton.in.us  574-223-2881   

                                           SITE_ADDR  
1  {'address1': '8343 State Route 19', 'address2'...  
2  {'address1': '216 NORTH 2ND STREET', 'address2...  
3  {'address1': '301 South Main Street, Suite 100...  
4  {'address1': '722 MAIN STREET', 'address2': na...  
5  {'address1': '125 EAST 9TH STREET', 'address2'...

In [0]:
agency_df.head()

AGENCY_ID                                  name  \
1         1  Tippecanoe Valley School Corporation   
2         2                   Guidance Ministries   
3         3    Elkhart County Prosecutor's office   
4         4           United Way of Fulton County   
5         5       Fulton County Health Department   

                                 AGENCY_AKA_KEYWORDS  \
1                                                NaN   
2                                                NaN   
3  Elhart County Victim's Assistance, Elkhart Cou...   
4                           Fulton County United Way   
5                                                NaN   

                                         description  \
1  Administers the elementary, middle, and high s...   
2  A nonprofit that provides a food and clothing ...   
3  Represents the state of Indiana in felony and ...   
4  A nonprofit agency that partners with local hu...   
5  A government agency that provides public healt...   

                                          AGENCY_HRS  \
1  {"open":{"Tuesday":[[[7,30],[16,30]]],"Monday"...   
2                                                NaN   
3                                                NaN   
4  {"open":{"Friday":[[[10,0],[15,0]]],"Thursday"...   
5  {"open":{"Friday":[[[8,0],[16,0]]],"Thursday":...   

                                      AGENCY_TXT_HRS  \
1                              Mon-Fri 7:30am-4:30pm   
2  Tues and Thurs 9am-6pm; Mon, Wed, and Fri 9am-4pm   
3  Mon 8am-noon and 1pm-5pm; Tues-Fri 8am-noon an...   
4                                   Mon-Fri 10am-3pm   
5                                    Mon-Fri 8am-4pm   

                    AGENCY_WEB_URL  \
1               www.tvsc.k12.in.us   
2       www.guidanceministries.com   
3  www.elkhartcountyprosecutor.com   
4                              NaN   
5              www.co.fulton.in.us   

                                       AGENCY_FB_URL AGENCY_TW_URL  \
1                                                NaN           NaN   
2                                                NaN           NaN   
3                                                NaN           NaN   
4  www.facebook.com/Fulton-County-United-Way-1444...           NaN   
5                                                NaN           NaN   

   AGENCY_PHONE AGENCY_PHONE_NOTE  
1  574-598-2759               NaN  
2  574-296-7192               NaN  
3  574-296-1888               NaN  
4  574-223-8929               NaN  
5  574-223-2881               NaN

In [0]:
service_df.head()

AGENCY_ID SERV_ID                   name SERV_AKA  \
1         1   21349        School District      NaN   
2         2    5245            Food Pantry      NaN   
3         2    6643          God's Kitchen      NaN   
4         2    6644       Computer Classes      NaN   
5         2    6645  Job Search Assistance      NaN   

                                         description  \
1  Administers the elementary, middle, and high s...   
2  Provides nonperishable food, meat, baby food, ...   
3  Offers a hot dinner twice weekly to all in nee...   
4  Offers free classes in computer skills to all ...   
5  Provides job search assistance to people needi...   

                                          SERV_SITES  \
1  [{"Site_Id":1,"Name":"Tippecanoe Valley School...   
2       [{"Site_Id":2,"Name":"Guidance Ministries"}]   
3       [{"Site_Id":2,"Name":"Guidance Ministries"}]   
4       [{"Site_Id":2,"Name":"Guidance Ministries"}]   
5       [{"Site_Id":2,"Name":"Guidance Ministries"}]   

                                        SERV_GEO_TAG  \
1  [{"scope":"county","county":"Benton","state":"...   
2  [{"scope":"county","county":"Elkhart","state":...   
3  [{"scope":"county","county":"Elkhart","state":...   
4  [{"scope":"county","county":"Elkhart","state":...   
5  [{"scope":"county","county":"Elkhart","state":...   

                                 SERV_TAXONOMY_CODES  \
1  [{"taxonomy_code":"HL-8120.8200"},{"taxonomy_c...   
2  [{"taxonomy_code":"BD-1800.2000"},{"taxonomy_c...   
3                 [{"taxonomy_code":"BD-5000.8300"}]   
4                 [{"taxonomy_code":"PL-7400.1500"}]   
5                 [{"taxonomy_code":"ND-3500.3600"}]   

                                       SERV_OP_HOURS SERV_CLOSED_TIMES  ...  \
1                              Mon-Fri 7:30am-4:30pm               NaN  ...   
2                                    Tues 9:30am-1pm               NaN  ...   
3                             Tues and Thurs 9am-6pm               NaN  ...   
4  Office: Tues and Thurs 9am-6pm; Mon, Wed, and ...               NaN  ...   
5  Office: Tues and Thurs 9am-6pm; Mon, Wed, and ...               NaN  ...   

  SERV_CLIENT_NEEDS_TO_BRING  \
1                       true   
2                       true   
3                      false   
4                      false   
5                      false   

                                       SERV_REQ_DOCS SERV_OPTS_BRING_LIST  \
1                                  Varies by service                  NaN   
2  Government-issued photo ID for all adults in h...                  NaN   
3                                     Nothing needed                  NaN   
4                                  Call for details.                  NaN   
5                         Call for more information.                  NaN   

  SERV_HAS_FEE SERV_FEE_STRUCT SERV_PAY_METHODS SERV_ALT_LANGS  \
1        false             NaN              NaN            NaN   
2        false             NaN              NaN            NaN   
3        false             NaN              NaN            NaN   
4        false             NaN              NaN            NaN   
5        false             NaN              NaN            NaN   

  SERV_WAIT_PERIOD SERV_CAP_AND_TYPE SERV_HELPLINE_NOTES  
1              NaN               NaN                 NaN  
2              NaN               NaN                 NaN  
3              NaN               NaN                 NaN  
4              NaN               NaN                 NaN  
5              NaN               NaN                 NaN  

[5 rows x 32 columns]

In [0]:
service_df.columns

Index(['AGENCY_ID', 'SERV_ID', 'name', 'SERV_AKA', 'description', 'SERV_SITES',
       'SERV_GEO_TAG', 'SERV_TAXONOMY_CODES', 'SERV_OP_HOURS',
       'SERV_CLOSED_TIMES', 'SERV_WEB_URL', 'SERV_EMAIL', 'SERV_PHONE_1',
       'SERV_PHONE_1_NOTE', 'SERV_PHONE_2', 'SERV_PHONE_2_NOTE',
       'SERV_PHONE_3', 'SERV_PHONE_3_NOTE', 'SERV_FAX', 'SERV_ELIGIBLE_DESCR',
       'SERV_HAS_PRE_REQ', 'SERV_INTAKE_PROC', 'SERV_CLIENT_NEEDS_TO_BRING',
       'SERV_REQ_DOCS', 'SERV_OPTS_BRING_LIST', 'SERV_HAS_FEE',
       'SERV_FEE_STRUCT', 'SERV_PAY_METHODS', 'SERV_ALT_LANGS',
       'SERV_WAIT_PERIOD', 'SERV_CAP_AND_TYPE', 'SERV_HELPLINE_NOTES'],
      dtype='object')

In [0]:
# save our datasets
# read here about data formats: https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html
# and here for simple comparison: https://towardsdatascience.com/the-best-format-to-save-pandas-data-414dca023e0d
# however, our data (particularly w. the json for location is not serializable for most libraries)

import numpy as np
agency_df.replace(np.nan, '', regex=True)
site_df.replace(np.nan, '', regex=True)
service_df.replace(np.nan, '', regex=True)

agency_df.to_csv('./data/nd_active_agency_dataset.csv', index=False)
service_df.to_csv('./data/nd_active_service_dataset.csv', index=False)
site_df.to_csv('./data/nd_active_site_dataset.csv', index=False)